File: pipeline/src/ixnos_data_pipeline/common/greek_text.py
"""Greek text normalisation for search. Must match GreekText.cs in the API: both run
contracts/text-normalisation/cases.json.

Postgres full-text already ignores accents; this is for trigram matching and Greeklish.
"""

import itertools
import re
import unicodedata

_WHITESPACE = re.compile(r"\s+")
_LETTER_RUN = re.compile(r"[^\W\d_]+")
_GREEK_LETTER = re.compile(r"[Ͱ-Ͽ]")
_LATIN_LETTER = re.compile(r"[a-z]")
_DOUBLED_LETTER = re.compile(r"([a-z])\1+")

# Lowercased Latin letters that look like Greek ones, mostly from uppercase data entry
# (A, B, E, H, I, K, M, N, O, P, T, X, Y, Z) plus lowercase v and u.
_HOMOGLYPHS = str.maketrans("abehikmnoptxyzvu", "αβεηικμνορτχυζνυ")

# Phonetic key tables. Two-letter entries are matched before one-letter ones.
_GREEK_KEYS = {
    "ου": "u", "ει": "i", "οι": "i", "υι": "i", "αι": "e", "αυ": "ab", "ευ": "eb",
    "μπ": "b", "ντ": "d", "γκ": "g", "γγ": "g",
    "α": "a", "β": "b", "γ": "g", "δ": "d", "ε": "e", "ζ": "z", "η": "i", "θ": "th",
    "ι": "i", "κ": "k", "λ": "l", "μ": "m", "ν": "n", "ξ": "ks", "ο": "o", "π": "p",
    "ρ": "r", "σ": "s", "τ": "t", "υ": "i", "φ": "f", "χ": "h", "ψ": "ps", "ω": "o",
}  # fmt: skip
_LATIN_KEYS = {
    "ou": "u", "ei": "i", "oi": "i", "yi": "i", "ui": "i", "ai": "e", "au": "ab",
    "av": "ab", "eu": "eb", "ev": "eb", "mp": "b", "nt": "d", "gk": "g", "gg": "g",
    "th": "th", "ch": "h", "kh": "h",
    "a": "a", "b": "b", "c": "k", "d": "d", "e": "e", "f": "f", "g": "g", "i": "i",
    "j": "tz", "k": "k", "l": "l", "m": "m", "n": "n", "o": "o", "p": "p", "q": "k",
    "r": "r", "s": "s", "t": "t", "u": "i", "v": "b", "w": "o", "y": "i", "z": "z",
}  # fmt: skip
# Greeklish letters with more than one reading, most likely first.
_AMBIGUOUS = {"h": ("i", "h"), "x": ("h", "ks")}
# Greeklish digits, only read as letters when next to a vowel ("8elo", "a3ia", not "3d").
_DIGIT_LETTERS = {"8": "th", "3": "ks"}
_LATIN_VOWELS = frozenset("aeiouyw")

MAX_EXPANDED = 3

# Case-preserving look-alike fix for display text (labels, names), where lowercasing is wrong.
_DISPLAY_HOMOGLYPHS = str.maketrans(
    "ABEHIKMNOPTXYZabehikmnoptxyzvu", "ΑΒΕΗΙΚΜΝΟΡΤΧΥΖαβεηικμνορτχυζνυ"
)
_ANY_LATIN_LETTER = re.compile(r"[A-Za-z]")


def normalise(text: str) -> str:
    """Lowercase, strip accents, fold final sigma, fix Latin look-alikes inside Greek
    words and collapse whitespace."""
    lowered = text.lower()
    decomposed = unicodedata.normalize("NFD", lowered)
    stripped = "".join(c for c in decomposed if unicodedata.category(c) != "Mn")
    folded = unicodedata.normalize("NFC", stripped).replace("ς", "σ").replace("µ", "μ")
    fixed = _LETTER_RUN.sub(_fix_homoglyphs, folded)
    return _WHITESPACE.sub(" ", fixed).strip()


def fix_latin_lookalikes(text: str) -> str:
    """Replaces Latin letters that look like Greek ones inside words that also contain Greek
    letters, keeping case: "Aττική" (Latin A) -> "Αττική". Words with no Greek letters are
    left alone. For display text; `normalise` does the same for search text."""

    def fix(match: re.Match[str]) -> str:
        word = match.group()
        if _GREEK_LETTER.search(word.lower()) and _ANY_LATIN_LETTER.search(word):
            return word.translate(_DISPLAY_HOMOGLYPHS)
        return word

    return _LETTER_RUN.sub(fix, text)


def search_keys(text: str) -> list[str]:
    """Each ambiguous Greeklish letter (h, x) doubles the keys, up to MAX_EXPANDED of them."""
    segments = _key_segments(normalise(text))
    ambiguous = [i for i, options in enumerate(segments) if len(options) > 1]
    expanded = set(ambiguous[:MAX_EXPANDED])
    choices = [options if i in expanded else options[:1] for i, options in enumerate(segments)]

    keys: list[str] = []
    for combination in itertools.product(*choices):
        key = _DOUBLED_LETTER.sub(r"\1", _WHITESPACE.sub(" ", "".join(combination)).strip())
        if key not in keys:
            keys.append(key)
    return keys


def search_key(text: str) -> str:
    """The key stored text is indexed under."""
    return search_keys(text)[0]


def _fix_homoglyphs(match: re.Match[str]) -> str:
    word = match.group()
    if _GREEK_LETTER.search(word) and _LATIN_LETTER.search(word):
        return word.translate(_HOMOGLYPHS)
    return word


def _key_segments(text: str) -> list[tuple[str, ...]]:
    segments: list[tuple[str, ...]] = []
    i = 0
    while i < len(text):
        pair, char = text[i : i + 2], text[i]
        if len(pair) == 2 and (pair in _GREEK_KEYS or pair in _LATIN_KEYS):
            segments.append((_GREEK_KEYS.get(pair) or _LATIN_KEYS[pair],))
            i += 2
            continue
        if char in _AMBIGUOUS:
            segments.append(_AMBIGUOUS[char])
        elif char in _GREEK_KEYS:
            segments.append((_GREEK_KEYS[char],))
        elif char in _LATIN_KEYS:
            segments.append((_LATIN_KEYS[char],))
        elif char in _DIGIT_LETTERS and _next_to_vowel(text, i):
            segments.append((_DIGIT_LETTERS[char],))
        elif char.isalnum():
            segments.append((char,))
        else:
            segments.append((" ",))
        i += 1
    return segments


def _next_to_vowel(text: str, i: int) -> bool:
    before = text[i - 1] if i > 0 else ""
    after = text[i + 1] if i + 1 < len(text) else ""
    return before in _LATIN_VOWELS or after in _LATIN_VOWELS

File: backend/src/IxnosData.Application/Search/GreekText.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace IxnosData.Application.Search;

// Keep in step with ixnos_data_pipeline.common.greek_text: both run
// contracts/text-normalisation/cases.json.
public static partial class GreekText
{
    // 2^3 = 8 keys at most.
    public const int MaxExpanded = 3;

    // Lowercased Latin letters that look like Greek ones, mostly from uppercase data entry.
    private const string HomoglyphsLatin = "abehikmnoptxyzvu";
    private const string HomoglyphsGreek = "αβεηικμνορτχυζνυ";

    private static readonly Dictionary<string, string> GreekKeys = new(StringComparer.Ordinal)
    {
        ["ου"] = "u",
        ["ει"] = "i",
        ["οι"] = "i",
        ["υι"] = "i",
        ["αι"] = "e",
        ["αυ"] = "ab",
        ["ευ"] = "eb",
        ["μπ"] = "b",
        ["ντ"] = "d",
        ["γκ"] = "g",
        ["γγ"] = "g",
        ["α"] = "a",
        ["β"] = "b",
        ["γ"] = "g",
        ["δ"] = "d",
        ["ε"] = "e",
        ["ζ"] = "z",
        ["η"] = "i",
        ["θ"] = "th",
        ["ι"] = "i",
        ["κ"] = "k",
        ["λ"] = "l",
        ["μ"] = "m",
        ["ν"] = "n",
        ["ξ"] = "ks",
        ["ο"] = "o",
        ["π"] = "p",
        ["ρ"] = "r",
        ["σ"] = "s",
        ["τ"] = "t",
        ["υ"] = "i",
        ["φ"] = "f",
        ["χ"] = "h",
        ["ψ"] = "ps",
        ["ω"] = "o",
    };

    private static readonly Dictionary<string, string> LatinKeys = new(StringComparer.Ordinal)
    {
        ["ou"] = "u",
        ["ei"] = "i",
        ["oi"] = "i",
        ["yi"] = "i",
        ["ui"] = "i",
        ["ai"] = "e",
        ["au"] = "ab",
        ["av"] = "ab",
        ["eu"] = "eb",
        ["ev"] = "eb",
        ["mp"] = "b",
        ["nt"] = "d",
        ["gk"] = "g",
        ["gg"] = "g",
        ["th"] = "th",
        ["ch"] = "h",
        ["kh"] = "h",
        ["a"] = "a",
        ["b"] = "b",
        ["c"] = "k",
        ["d"] = "d",
        ["e"] = "e",
        ["f"] = "f",
        ["g"] = "g",
        ["i"] = "i",
        ["j"] = "tz",
        ["k"] = "k",
        ["l"] = "l",
        ["m"] = "m",
        ["n"] = "n",
        ["o"] = "o",
        ["p"] = "p",
        ["q"] = "k",
        ["r"] = "r",
        ["s"] = "s",
        ["t"] = "t",
        ["u"] = "i",
        ["v"] = "b",
        ["w"] = "o",
        ["y"] = "i",
        ["z"] = "z",
    };

    // Greeklish letters with more than one reading, most likely first.
    private static readonly Dictionary<char, string[]> Ambiguous = new()
    {
        ['h'] = ["i", "h"],
        ['x'] = ["h", "ks"],
    };

    // Greeklish digits, read as letters only next to a vowel ("8elo", "a3ia", not "3d").
    private static readonly Dictionary<char, string> DigitLetters = new() { ['8'] = "th", ['3'] = "ks" };

    private const string LatinVowels = "aeiouyw";

    // Lowercase, no accents, final sigma folded, Latin look-alikes fixed, whitespace collapsed.
    public static string Normalise(string text)
    {
        // Lowercase (not uppercase) to match the Python side and PostgreSQL lower().
#pragma warning disable CA1308
        var decomposed = text.ToLowerInvariant().Normalize(NormalizationForm.FormD);
#pragma warning restore CA1308
        var stripped = new StringBuilder(decomposed.Length);
        foreach (var c in decomposed)
        {
            if (CharUnicodeInfo.GetUnicodeCategory(c) != UnicodeCategory.NonSpacingMark)
            {
                stripped.Append(c);
            }
        }

        var folded = stripped.ToString()
            .Normalize(NormalizationForm.FormC)
            .Replace('ς', 'σ')
            .Replace('\u00b5', '\u03bc'); // micro sign -> Greek mu (they look identical)
        var fixedText = LetterRun().Replace(folded, FixHomoglyphs);
        return Whitespace().Replace(fixedText, " ").Trim();
    }

    // Each ambiguous Greeklish letter (h, x) doubles the keys, up to MaxExpanded of them.
    public static IReadOnlyList<string> SearchKeys(string text)
    {
        var segments = KeySegments(Normalise(text));

        var expandedLeft = MaxExpanded;
        var partials = new List<string> { string.Empty };
        foreach (var options in segments)
        {
            var choices = options;
            if (options.Length > 1)
            {
                choices = expandedLeft > 0 ? options : [options[0]];
                expandedLeft--;
            }

            partials = [.. partials.SelectMany(prefix => choices.Select(choice => prefix + choice))];
        }

        return [.. partials
            .Select(key => DoubledLetter().Replace(Whitespace().Replace(key, " ").Trim(), "$1"))
            .Distinct(StringComparer.Ordinal)];
    }

    // The key stored text is indexed under.
    public static string SearchKey(string text) => SearchKeys(text)[0];

    private static string FixHomoglyphs(Match match)
    {
        var word = match.Value;
        if (!GreekLetter().IsMatch(word) || !LatinLetter().IsMatch(word))
        {
            return word;
        }

        var chars = word.ToCharArray();
        for (var i = 0; i < chars.Length; i++)
        {
            var index = HomoglyphsLatin.IndexOf(chars[i], StringComparison.Ordinal);
            if (index >= 0)
            {
                chars[i] = HomoglyphsGreek[index];
            }
        }

        return new string(chars);
    }

    private static List<string[]> KeySegments(string text)
    {
        var segments = new List<string[]>(text.Length);
        var i = 0;
        while (i < text.Length)
        {
            if (i + 1 < text.Length)
            {
                var pair = text.Substring(i, 2);
                if (GreekKeys.TryGetValue(pair, out var pairKey) || LatinKeys.TryGetValue(pair, out pairKey))
                {
                    segments.Add([pairKey]);
                    i += 2;
                    continue;
                }
            }

            var c = text[i];
            var single = c.ToString();
            if (Ambiguous.TryGetValue(c, out var readings))
            {
                segments.Add(readings);
            }
            else if (GreekKeys.TryGetValue(single, out var key) || LatinKeys.TryGetValue(single, out key))
            {
                segments.Add([key]);
            }
            else if (DigitLetters.TryGetValue(c, out var digitKey) && NextToVowel(text, i))
            {
                segments.Add([digitKey]);
            }
            else
            {
                segments.Add([char.IsLetterOrDigit(c) ? single : " "]);
            }

            i++;
        }

        return segments;
    }

    private static bool NextToVowel(string text, int i) =>
        (i > 0 && LatinVowels.Contains(text[i - 1], StringComparison.Ordinal))
        || (i + 1 < text.Length && LatinVowels.Contains(text[i + 1], StringComparison.Ordinal));

    [GeneratedRegex(@"\s+")]
    private static partial Regex Whitespace();

    [GeneratedRegex(@"\p{L}+")]
    private static partial Regex LetterRun();

    [GeneratedRegex(@"[Ͱ-Ͽ]")]
    private static partial Regex GreekLetter();

    [GeneratedRegex("[a-z]")]
    private static partial Regex LatinLetter();

    [GeneratedRegex("([a-z])\\1+")]
    private static partial Regex DoubledLetter();
}
